manageiq-external-auth: do NOT require authentication for /api/product_info

himdel opened the pull request:
The API has 2 sets of requests which can go in unauthenticated:

    app/controllers/api/base_controller.rb
    25:    before_action :require_api_user_or_token, :except => [:options, :product_info]
    27:    before_action :validate_api_request, :except => [:product_info]
    28:    before_action :validate_api_action, :except => [:options, :product_info]

That's OPTIONS requests, and a GET for `/api/product_info`.

We don't need OPTIONS quite yet (except for CORS preflight, which is not supported for the API currently),

but the Service UI is relying on unauthenticated `/api/product_info` to retrieve the URLs for brand assets (like the login screen logo).

Without this, the SUI login screen shows a browser login popup (and no branding), when running with external authentication.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1790817

Cc @jvlcek , thanks for all the help! :)

himdel committed:
manageiq-external-auth: do NOT require authentication for /api/product_info

API has 2 sets of requests which can go in unauthenticated:

    app/controllers/api/base_controller.rb
    25:    before_action :require_api_user_or_token, :except => [:options, :product_info]
    27:    before_action :validate_api_request, :except => [:product_info]
    28:    before_action :validate_api_action, :except => [:options, :product_info]

That's OPTIONS requests, and a GET for /api/product_info.

We don't need OPTIONS quite yet (except for CORS preflight, which is not supported for the API currently),

but the Service UI is relying on unauthenticated /api/product_info to retrieve the URLs for brand assets (like the login screen logo).

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1790817

jvlcek commented:
👍 Looks good to me!

himdel commented:
An alternative may be something like...

`SetEnvIf Request_URI ^/api/product_info$  let_unauthenticated_in`
...
`Allow from env=let_unauthenticated_in`

---

(Then we could add `SetEnvIf Request_Method OPTIONS let_unauthenticated_in`, but we may need to talk to the actual API to know which OPTIONS requests should go in and which do need to be authenticated.)

skateman commented:
Yeah, `product_info` should go without any authentication as the SUI uses it to render the login screen.

carbonin commented:
@abellotti can you take a look at this one?

abellotti commented on TEMPLATE/etc/httpd/conf.d/manageiq-external-auth.conf.erb:
```diff
@@ -37,7 +37,7 @@ LoadModule auth_kerb_module modules/mod_auth_kerb.so
 
 
-<LocationMatch ^/api|^/vmdbws/wsdl|^/vmdbws/api>
+<LocationMatch ^/api(?!\/product_info$)|^/vmdbws/wsdl|^/vmdbws/api>
```
this should match the optional versioned path, i.e. /api/v4.1.0/product_info, as per https://github.com/ManageIQ/manageiq-api/blob/086814a4a80cd1e83e2125338f96859d94394b20/lib/api.rb#L3, though probably not used by the SUI, but for completeness if anything.

At some point, we should also remove the /vmdbws/* entries, (the SOAP interface), those are LONG gone.

himdel replied:
OK, removing vmdbws.

And looks like `^\/api(?!\/(v[\d\.]+\/)?product_info$)` is the right negative lookahead to handle an optional `v + numbers and dots`

himdel replied:
(@abellotti That's the only format for the api versions, right? `v1` `v1.2.3` etc.)

himdel committed:
manageiq-external-auth: drop vmdbws, support /api/v1.3.4/ when matching product_info

/api/something - need auth
/api/product_info - no auth
/api/v1.2.3/product_info - no auth
/api/product_info/foo - need auth

himdel commented:
@miq-bot add_label ivanchuk/yes, hammer/yes

SUI is using the request since https://github.com/ManageIQ/manageiq-ui-service/pull/1513, that's hammer/backported.

simaishi commented:
Ivanchuk backport details:

```
$ git log -1
commit fe254e3689727ca2cb5be3ac7d238270376578a8
Author: Brandon Dunne <bdunne@redhat.com>
Date:   Tue Feb 25 12:29:42 2020 -0500

    Merge pull request #272 from himdel/bz1790817

    manageiq-external-auth: do NOT require authentication for /api/product_info
    (cherry picked from commit 739ae404f7559d3e63d5bd759fad90f68ab99c05)

    https://bugzilla.redhat.com/show_bug.cgi?id=1790817
```
